File: week-03_05/tp/tp_py/tp2/ping_pong_3.py
from mpi4py import MPI
import numpy as np

# Parameters
nb_values_max = 7000000
nb_tests = 10
tag = 99

# TODO: Get rank -> rank


# Values array
values = np.zeros(nb_values_max, dtype=np.float64)

# List of messages sizes to test
nb_values = [0, 0, 1, 10, 100, 1000, 10000, 100000, 1000000, 7000000]

for nb_value in nb_values:
    # Do a ping for each size

        values = np.random.rand(nb_values_max)
        # TODO: Begin the timing of the ping pong -> time_begin

        # TODO: Process 0: Send and recv with process 1

        # TODO: End the timing of the ping pong -> time_end

        bandwidth = (2 * nb_value * 8) / 1000000 / (time_end - time_begin)
        if nb_value != 0:
            print(f"Me, process 0, sent and received {nb_value} "
                  f"values (last = {values[nb_value - 1]:.2f}) "
                  f"from process 1 in {time_end - time_begin:.6f} seconds,"
                  f" bandwidth {bandwidth:.2f} Mo/s.")
        else:
            print(f"Me, process 0, sent and received {nb_value} "
                  f"values in {time_end - time_begin:.6f} seconds, "
                  f"bandwidth {bandwidth:.2f} Mo/s.")

        # TODO: Process 1: Recv and send with process 0

File: week-03_05/tp/tp/tp2/ping_pong_3.f90
program ping_pong_3
  use, intrinsic :: iso_fortran_env, only : dp => real64
  USE MPI_F08
  implicit none

  TYPE(MPI_Status)                             :: status
  integer, parameter                           :: nb_values_max=7000000, &
                                                  nb_tests=10,tag=99
  integer, dimension(nb_tests)                 :: nb_values
  integer                                      :: rank,i
  real(kind=dp), dimension(0:nb_values_max-1)  :: values
  TYPE(MPI_Datatype)                           :: typedp
  real(kind=dp)                                :: time_begin,time_end

  ! TODO: Get rank -> rank

  nb_values = (/ 0,0,1,10,100,1000,10000,100000,1000000,7000000 /)

  do i=1,nb_tests
    if (rank == 0) then
      call random_number(values)
      ! TODO: Start timing -> time_begin

      ! TODO: Process 0 send values to and receive from process 1

      ! TODO: End timing -> time_end

      call print_function
    elseif (rank == 1) then
      ! TODO: Process 1 receive values from and send to process 0

    end if
  end do

  ! TODO: Finalize MPI
  
contains
  subroutine print_function
    if (nb_values(i)/=0) then
      print ('("Me, process 0, sent and received ",i8, &
           & " values (last = ",f4.2,") from process 1", &
           & " in  ",f8.6," seconds, bandwidth ",f7.2, &
           & " Mo/s.")'), &
           nb_values(i),values(nb_values(i)-1),time_end-time_begin, &
           real(2*nb_values(i)*8)/1000000./(time_end-time_begin)
    else
      print ('("Me, process 0, sent and received ",i8, &
           & " values in ",f8.6," seconds, bandwidth ",f7.2, &
           & " Mo/s.")'), &
           nb_values(i),time_end-time_begin, &
           real(2*nb_values(i)*8)/1000000./(time_end-time_begin)
    end if
  end subroutine print_function
end program ping_pong_3
